### PR TITLE
Event interfaces

### PR DESCRIPTION
**Pull request by mmihic**

### CODE REVIEW

File: misk-events/src/main/kotlin/misk/events/Event.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  /**
+   * Partitioning key for the event. The partitioning key controls the ordering and sharding
+   * of events on a topic. Events on a topic with the same partitioning key are delivered on
+   * the same shard and in the order in which they were published.  Typically, the entity
+   * identifier is also used as a partitioning key, such that all of events on a topic for
+   * a given entity get delivered in the order in which they were submitted. However,
+   * producing applications may include an alternate partition key as part of the event
+   * to support ordering/sharding at a level above the individual entity. For example,
+   * a card processing system may want to shard and order all credit card changes relative
+   * to the customer to whom the card belongs; in this case the entity identifier is the credit
```
Comment (ryanhall07) [inline misk-events/src/main/kotlin/misk/events/Event.kt]: what is the value of the `entityIdentifier`?

File: misk-events/src/main/kotlin/misk/events/SpooledProducer.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package misk.events
+
+import misk.hibernate.Gid
+import misk.hibernate.Session
+
+/**
+ * A [SpooledProducer] is a producer that writes events to a local spool stored within a
+ * service's database. [SpooledProducer]s can be used to
```
Comment (ryanhall07) [inline misk-events/src/main/kotlin/misk/events/SpooledProducer.kt]: finish the thought?

File: misk-events/src/main/kotlin/misk/events/SpooledProducer.kt [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package misk.events
+
+import misk.hibernate.Gid
+import misk.hibernate.Session
+
+/**
+ * A [SpooledProducer] is a producer that writes events to a local spool stored within a
```
Comment (ryanhall07) [inline misk-events/src/main/kotlin/misk/events/SpooledProducer.kt]: so if I want to skip the db tx, I would use a `Producer` directly?

File: misk-events/src/main/kotlin/misk/events/Producer.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  /**
+   * A producer [Transaction] is a unit of publishing. Producer transactions adhere to the
+   * transactional concepts of isolation (events publishing within a transaction are only
+   * visible to consumers once the transaction is committed) and atomicity (once committed, all
+   * events within the transactions are made available to consumers - no events will be lost).
+   *
+   * Note that a Producer transaction is _not_ connected to a database transaction; it is solely
+   * a transaction within the event streaming system. To coordinate event publishing with
+   * local database transactions, use a [SpooledProducer] which stores the events in a local table
+   * as part of the local database transaction
```
Comment (wesleyk) [inline misk-events/src/main/kotlin/misk/events/Producer.kt]: worthwhile to document behavior of long-lived transactions? Do transactions ever time out/auto-rollback?

File: misk-events/src/main/kotlin/misk/events/SpooledProducer.kt (hunk 2 above)
Comment (mmihic) [reply misk-events/src/main/kotlin/misk/events/SpooledProducer.kt]: oops sorry, done

File: misk-events/src/main/kotlin/misk/events/Producer.kt (hunk 4 above)
Comment (mmihic) [reply misk-events/src/main/kotlin/misk/events/Producer.kt]: Sure

**Pull request merged**